### PR TITLE
recreate fixes

### LINKED ISSUE #1920: rechunk lost all file content
Note: it looks like this is a bug in (unreleased) borg 1.1, borg 1.0.x is not affected.

borg 1.1 (master branch), while trying to reproduce #1658, all went apparently(!) well, all checks succeeded, no tracebacks, no error msgs.

but then I noticed that in an archive list, all file lengths are 0.
when extracting the file, a 0 bytes file was extracted.
other metadata looked ok (uid/gid/timestamp/path).

what i roughly did was:

- copied bigger production repo to usb3 disk
- borg check --repository-only
- borg recreate (changed compression) on some archives
- borg delete on bigger number of archives
- borg prune, reducing the whole repo to ~ 10 archives
- borg check (complete, archives too)
- borg recreate with --exclude on a bigger directory
- borg check (complete, archives too)
- borg check --verify-data

I'll try to reproduce with lesser actions.

### PR DESCRIPTION
**Pull request by enkore**
Fixes #1920 #1919 #1913

#1920 was introduced in 93b03ea2310. Previously the function always returned an iterator (a3ee9d2c). In 93b03ea I rewrote it to *be* the iterator, without also modifying the second code path to conform. This was not noticed because the test for this code path was broken, and because I did not take a closer look at the resulting archives during manual testing. Just using `recreate ... --stats` would have already shown that something was amiss.

#1919 was overlooked when removing the signal processing stuff. Archiver.do_recreate used the return value to exit should a signal be received; these returns where removed, but the if in do_recreate was not updated.

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
@enkore looks promising - i presume for a repository affected by this i still need to run the code in the gist you posted?

**Comment by enkore** [comment]
This doesn't change anything about the other bug (#1658) you experience. The gist only fixes #1598.

**Comment by RonnyPfannschmidt** [comment]
@enkore i see

File: src/borg/archive.py [hunk 1]
```diff
@@ -1517,24 +1516,22 @@ def chunk_processor(self, target, compress, data):
...
-    def create_chunk_iterator(self, archive, target, chunks):
-        """Return iterator of chunks to store for 'item' from 'archive' in 'target'."""
+    def iter_chunks(self, archive, target, chunks):
         chunk_iterator = archive.pipeline.fetch_many([chunk_id for chunk_id, _, _ in chunks])
         if target.recreate_rechunkify:
             # The target.chunker will read the file contents through ChunkIteratorFileWrapper chunk-by-chunk
             # (does not load the entire file into memory)
             file = ChunkIteratorFileWrapper(chunk_iterator)
-            return target.chunker.chunkify(file)
+            yield from target.chunker.chunkify(file)
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: interesting.

but regarding the 2 lines below in the else branch, maybe rather use:

```
for chunk_data in target.chunker.chunkify(file):
    yield chunk_data
```

File: src/borg/archiver.py [hunk 2]
```diff
@@ -2387,8 +2388,8 @@ def build_parser(self, prog=None):
 
         When rechunking space usage can be substantial, expect at least the entire
         deduplicated size of the archives using the previous chunker params.
-        When recompressing approximately 1 % of the repository size or 512 MB
-        (whichever is greater) of additional space is used.
+        When recompressing expect approx throughput / checkpoint-interval in space usage,
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: add a dot when abbreviating.
also maybe (t / c-i) makes it clearer that it is meant as a computation.

**Comment by ThomasWaldmann** [comment]
would like to merge this soon, to do more testing on this code.

**Pull request merged**